EIPS/eip-7937.md

```diff
@@ -0,0 +1,145 @@
+discussions-to: https://ethereum-magicians.org/t/eip-9687-64-bit-mode-evm-operations/23794
+status: Draft
+type: Standards Track
+category: Core
+created: 2025-04-23
+---
+
+## Abstract
+
+This EIP introduces multibyte opcodes prefixed by `C0` for 64-bit arithmetic (`C001`-`C00B`), comparison (`C010`-`C015`), bitwise (`C016`-`C019`), stack (`DUP*` and `SWAP*`) and flow (`C056` and `C057` in "legacy", or `C0E1` and `C0E2` in EOFv1) operations.
+
+## Motivation
+
+Not all computations in EVM can utilize the full 256-bit integer width. It can therefore be beneficial to have a "64-bit mode" to avoid unnecessary cycles. This EIP uses a "prefix" opcode `C0`, essentially forming multibyte opcodes to avoid polluting the EVM opcode space too much.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Prefix opcode behavior
+
+This EIP uses the prefix opcode `C0`, and it only occupies this single EVM opcode space. Upon the interpreter encountering opcode `C0`, it MUST continue to seek the next byte in code. It then executes things in "64-bit mode", based on the second byte, described below. If the execution is successful, then the interpreter MUST increase `PC` by 2 (instead of 1).
+
+If the second byte is not a valid 64-bit mode operation, then the interpreter MUST OOG.
+
+### General 64-bit mode behavior
+
+In 64-bit mode, all operations only work on the lowest 64-bit of each stack value. The first 192-bit is discarded. When a result value is pushed back onto the stack, then it MUST ensures that observable effects will see that the first 192-bit is zero. Note that here it's not necessary for an interpreter to reset the first 192-bit to zero every time -- if the next opcode is still in 64-bit mode, then the first 192-bit is still unobservable. We discuss the full details in the "rationale" section. The interpreter only needs to reproduce the full 256-bit value upon entering non-64-bit mode. If the full computational heavy part can be written in pure 64-bit mode, then this can result in noticable performance gain.
+
+### Gas cost constants
+
+We define the following gas cost constants:
+
+* `G_BASE64`: 1
+* `G_VERYLOW64`: 2
+* `G_LOW64`: 3
+* `G_MID64`: 4
+
+### Arithmetic opcodes
+
+The 64-bit mode arithmetic opcodes are defined the same as non-64-bit mode, except that it only operates on the last 64-bits. In the below definition, `a`, `b`, `N` is `a mod 2^64`, `b mod 2^64` and `N mod 2^64`.
+
+* ADD (`C001`) and SUB (`C003`): `a op b mod 2^64`, gas cost `G_VERYLOW64`.
+* MUL (`C002`), DIV (`C004`), SDIV (`C005`), MOD (`C006`), SMOD (`C007`), SIGNEXTEND (`C00B`): `a op b mod 2^64`, gas cost `G_LOW64`.
+* ADDMOD (`C008`), MULMOD (`C009`): `a op b % N mod 2^64`, gas cost `G_MID64`.
+* EXP (`C00A`): `a EXP b mod 2^64`, gas cost is the same as non-64-bit mode.
```

Comment:
The EXP gas costs are dynamic: it depends on the exponent, and is not a standard gas cost (not sure if this is meant by this sentence to just copy that gas cost)

Reply:
Yes, great approach, usually EIPs are underspecified instead of overspecified so this approach would be great. For the per-opcode specification can't you "just" run the same opcode as the 256-bit variant, but before the opcode runs you take the relevant stack items and `mod 2^64` them to remove the topmost bits. This is now your input to the opcode. Now you evaluate the opcode, and for the output you have to `mod 2^64` it again. What rests is to add the original "topmost bits" (so topmost 192 bits) back into the stack item (I think you can just ADD / XOR the topmost bits in)